File: testing/CS3310fall2014/3310website/cs3310/homepages.wmich.edu/_kaminski/3310/PrettyPrintUtility.java
/**
 * Name:    Joshua White
 * Section: CS3310-100
 * Package: Countries of the World v.1.5
 * Desc:    PrettyPrintUtility class, v.1.5.4. Reads country data
 * 	    from specified file and displays contents in 
 * 	    log file. Use of this utility requires the following:
 * 
 * 	    1. Data records must be exactly 82 bytes long, written in this format
 *  	       (per assignment specs):
 *  	       - 6 bytes for countryCode (3 chars in default encoding - 
 *  		 UTF-16 - 2 bytes each)
 * 	       - 2 bytes for countryID (written as a short)
 * 	       - 30 bytes for name (15 chars in default encoding)
 * 	       - 26 bytes for continent (13 chars in default encoding)
 * 	       - 4 bytes for area (written as an int)
 * 	       - 8 bytes for population (written as a long)
 * 	       - 4 bytes for life expectancy (written as a single-precision float)
 * 	       - 2 bytes for link (written as a short)
 *  	    2. MAX_N_HOME_LOC, nHome, and nColl must be written in that
 *  	       order, beginning at byte 0
 *	    3. The header must be exactly 6 bytes long (ie., the first data
 *	       record begins at byte 6)
 *
 *	If you have any issues implementing this in your program, feel free
 *	to contact me at joshua.j45.white@wmich.edu. Please put "CS3310" in the
 *	subject line. Before emailing me, please make sure your program
 *	meets the specifications listed above, and that you have reviewed all
 *	notes in the code below.
 *
 *	This utility has been tested and works properly in my own program.
 * 
 *		This code is written in compliance with JavaSE-1.7
 */
package edu.wmich.cs3310.jwhite_cotw;

import java.io.BufferedWriter;
import java.io.DataInputStream;
import java.io.EOFException;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.nio.channels.FileChannel;
import java.text.NumberFormat;

public class PrettyPrintUtility {

	/* ******************************** *
	 *  	      Main method	    *
	 * ******************************** */

	/**
	 * Main method.
	 * @param args command line arguments
	 * @throws IOException on I/O error
	 */
	public static void main(String[] args) throws IOException {

	 	// Local variables
		NumberFormat nf = NumberFormat.getInstance();
		FileInputStream fstream;

	 	// Try with resources - ensures that resources declared are closed
	 	// properly after try block finishes executing (requires declared 
	 	// resources implement the java.io.AutoCloseable interface)
		try (
			DataInputStream file = new DataInputStream(
				fstream = new FileInputStream(new File("CountryData.bin")));
			PrintWriter log = new PrintWriter(new BufferedWriter(
				new FileWriter("log.txt", true)));
			) {
			
		 	// Write initialization messages to log
			log.println("STATUS > Log FILE opened");
			log.println("STATUS > CountryData FILE opened");
			
		 	// File channel from file stream
			FileChannel fc = fstream.getChannel();
			
		 	// Variables for record data
			char[] continent, countryCode, name;
			short max, nColl, nHome, countryID, link;
			int area, loc = 1;
			long population;
			float lifeExp;
		
			// End of file flag
			boolean eof = false;
			
		 	// Read header record data
			fc.position(0);
			max = file.readShort();
			nHome = file.readShort();
			nColl = file.readShort();
			
		 	// Write data table header in log
			log.println("DATA STORAGE");
			log.printf("MAX_N_HOME_LOC: %02d, nHome: %02d, nColl: %02d%n",
					max, nHome, nColl); 
			log.println("LOC/ CDE ID- NAME----------- CONTINENT---- "
					+ "------AREA ---POPULATION LIFE LINK");
			
			// Loop until end of file flag is set
			while (!eof) {

			 	// DataInputStream throws EOFException at end of file, so wrap
			 	// any stream methods in try block
				try {

				 	// Move file channel to next record location
					fc.position(offset(loc));

				 	/* Read country code (note algorithm used -- 	*
				 	 * ensure your write method follows this same 	*
				 	 * algorithm (this is the algorithm is used to	*
				  	 * read the name and continent fields and is as	*
				 	 * specified in the assignment specs)		*/
					countryCode = new char[3];
					for (int i = 0; i < countryCode.length; i++)
						countryCode[i] = file.readChar();
					
				 	// Read country ID
					countryID = file.readShort();
					
				 	// Read name
					name = new char[15];
					for (int i = 0; i < name.length; i++)
						name[i] = file.readChar();
					
				 	// Read continent
					continent = new char[13];
					for (int i = 0; i < continent.length; i++)
						continent[i] = file.readChar();
					
				 	// Read area, population, life expectancy, and link
					area = file.readInt();
					population = file.readLong();
					lifeExp = file.readFloat();
					link = file.readShort();

				 	// Execute if country ID is not all zero bits - ie., a valid record
					if (countryID != 0x00) 
						log.printf("%03d/ %3s %03d %-15.15s %-13.13s"
						+ " %10s %13s %4.1f %03d%n", loc, 
						new String(countryCode), countryID, 
						new String(name), new String(continent), 
						nf.format(area), nf.format(population),
						lifeExp, link);
					
				 	// Otherwise country ID is zero, which identifies and invalid record
					else
						log.printf("%03d/%n", loc);

				 	// Increment data location counter
					loc++;

			 	// When EOF is reached
				} catch (EOFException e) {

				 	// Set EOF flag
					eof = true;

				}
				
			}
			
		 	// Write data table footer
			log.println("++++++++++++++++++++++++++++++++++++++++++"
					+ "+++++++++++++++++++++++++++++++++++");
			log.println();

		 	// Write closing messages to log
			log.println("STATUS > CountryData FILE closed");
			log.println("STATUS > Log FILE closed");

		} catch (IOException e) {

			e.printStackTrace();

		}

	}

	/* ******************************** *
	 * 		Methods		    *
	 * ******************************** */

	/**
	 * Calculates byte offset.
	 * @param i storage location to calculate
	 * @return byte offset
	 */
	private static long offset(int i) {

		return ((i - 1) * 82) + 6;
		
	}

}

File: testing/CS3310fall2014/3310website/cs3310/homepages.wmich.edu/_kaminski/3310/PrettyPrintUtility.cs
/* Project:         Countries of the World
 * Module:          PrintUtility
 * Programmer:      Dan Peekstok
 *
 * This module prints the contents of the CountryData file to the Log file.
 * It is written without OOP because of the simplisity of the code
 * 
 * 
 * If you have any questions, Email me at dpeekstok@gmail.com or daniel.g.peekstok@wmich.edu
 * Please have A3 in the subject line. I will be more likely to respond to the first
 * email address. I dont check WMU as often as I should.
 * 
 * updated 10/14/14 at 11:30
 */


using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;

namespace PrintUtility
{
    public class PrintUtility
    {
            //private variables

        //holds the data of each country as it is read in
        private static string[] countryArray = new string[4];

        //private streams used for input.
        private static FileStream dataFileStream;
        private static BinaryReader countryDataReader;


        public static void Main()
        {
            Console.WriteLine("print");
            StreamWriter logWriter;

            //truncates the log if it hasn't been done yet since the start of the program
            //if it has been done, it just opens it in append mode
            
            logWriter = File.AppendText("Log.txt");
            logWriter.WriteLine("STATUS > Log File opened");

            logWriter.WriteLine("STATUS > PrintUtility started");

            countryDataReader = new BinaryReader(File.Open("CountryData.bin", FileMode.Open));

            //since strings are prefixed with the length, there is 1 added to the size of al the strings
            int sizeOfCode = 3, sizeOfId = sizeof(short), sizeOfName = 15, sizeOfCont = 13, sizeOfArea = sizeof(int), 
            sizeOfPop = sizeof(long), sizeOfLE = sizeof(float), sizeOfPointer = sizeof(short);
            int sizeOfHeader = sizeof(short) + sizeof(short) + sizeof(short);

            //variables used for output
            string code;
            string cont;
            int sa;
            long pop;
            float le;
            short id;
            string name;
            short pointer;

            int sizeOfRecord = sizeOfCode + sizeOfId + sizeOfName + sizeOfCont + sizeOfArea + 
                sizeOfPop + sizeOfLE + sizeOfPointer;

            short maxHome = countryDataReader.ReadInt16();
            short nHome = countryDataReader.ReadInt16();
            short nColl = countryDataReader.ReadInt16();
            int numOfCountries = nHome + nColl;
            
            //header
            logWriter.WriteLine("Max Home Locations: " + maxHome);
            logWriter.WriteLine("Number in Home Locations: " + nHome);
            logWriter.WriteLine("Number of Collisions: " + nColl);
            logWriter.WriteLine("LOC/ CDE ID- NAME----------- CONTINENT---- ------AREA " +
                "---POPULATION LIFE LINK");    //header

            //this will do a linear search of the entire file. 
            for (int x = 1; x <= maxHome + nColl; x++)
            {
                try
                {
                    //since the BinaryReader type doesn't have a Seek function, I used this work around. 
                    //its kinds sloppy but it works perfectly
                    //If the file location is empty, it wont advance the reader to the correct spot.

                    //first you close the old reader
                    countryDataReader.Close();
                    //then you create a file stream and use its seek function
                    dataFileStream = new FileStream("CountryData.bin", FileMode.Open);
                    dataFileStream.Seek(((x - 1) * sizeOfRecord) + sizeOfHeader, SeekOrigin.Begin);
                    //when you create a new BinaryReader on the same FileStream, the pointer will be in the same spot
                    countryDataReader = new BinaryReader(dataFileStream);


                    //I output the data to my CountryData.bin file in exactly this manner.
                    //If you use the same format of output, you wont have a problem
                    code = new string(countryDataReader.ReadChars(3));
                    id = countryDataReader.ReadInt16();
                    name = new string(countryDataReader.ReadChars(15));
                    cont = new string(countryDataReader.ReadChars(13));
                    sa = countryDataReader.ReadInt32();
                    pop = countryDataReader.ReadInt64();
                    le = countryDataReader.ReadSingle();
                    pointer = countryDataReader.ReadInt16();
                    

                    if (id == 0)
                        logWriter.WriteLine(x.ToString("000") + "/ Empty");
                    else
                    {

                        logWriter.WriteLine(String.Format("{0,3:000}/ {1,3} {2:000} {3, -15} " +
                            "{4,-13} {5,10:#,##0} {6,13:#,##0} {7,4:#0.0}  {8,2:#0}",
                            x, code, id, name, cont, sa, pop, le, pointer));
                    }
                }
                catch
                {
                    logWriter.WriteLine(x.ToString("000") + "/ Empty");
                }
            }

            
            logWriter.WriteLine("STATUS > PrintUtility finished");

            logWriter.WriteLine("STATUS > Backup File closed");

            countryDataReader.Close();
            logWriter.WriteLine("STATUS > Log File closed");
            logWriter.Close();

        }
    }
}
